packages/instant/src/components/buy_button.tsx

```diff
@@ -35,21 +36,23 @@ export class BuyButton extends React.Component<BuyButtonProps> {
     }
     private readonly _handleClick = async () => {
         // The button is disabled when there is no buy quote anyway.
-        if (_.isUndefined(this.props.buyQuote) || _.isUndefined(this.props.assetBuyer)) {
+        const { buyQuote, assetBuyer } = this.props;
+        if (_.isUndefined(buyQuote) || _.isUndefined(assetBuyer)) {
             return;
         }
-        this.props.onClick(this.props.buyQuote);
+        this.props.onAwaitingSignature(buyQuote);
         let txnHash;
         try {
-            txnHash = await this.props.assetBuyer.executeBuyQuoteAsync(this.props.buyQuote);
-            const txnReceipt = await web3Wrapper.awaitTransactionSuccessAsync(txnHash);
-            this.props.onBuySuccess(this.props.buyQuote, txnReceipt.transactionHash);
+            txnHash = await assetBuyer.executeBuyQuoteAsync(buyQuote);
```

Comment:
If we split this up into multiple try/catches i believe that will give us a bit more control over handling the errors, it will also let us get rid of the optional txHash in the `onBuyFailure` callback:

```
    private readonly _handleClick = async () => {
        // The button is disabled when there is no buy quote anyway.
        const { buyQuote, assetBuyer } = this.props;
        if (_.isUndefined(buyQuote) || _.isUndefined(assetBuyer)) {
            return;
        }
        let txnHash;
        this.props.onAwaitingSignature(buyQuote);
        try {
            txnHash = await assetBuyer.executeBuyQuoteAsync(buyQuote);
        } catch (e) {
            if (e instanceof Error && e.message === AssetBuyerError.SignatureRequestDenied) {
                this.props.onSignatureDenied(buyQuote);
                return;
            }
            throw e;
        }
        this.props.onBuyProcessing(buyQuote, txnHash);
        try {
            await web3Wrapper.awaitTransactionSuccessAsync(txnHash);
        } catch (e) {
            if (e instanceof Error && e.message.startsWith('Transaction failed')) {
                this.props.onBuyFailure(buyQuote, txnHash);
                return;
            }
            throw e;
        }
        this.props.onBuySuccess(buyQuote, txnHash);
    };
```

Reply:
This implementation also re-throws errors that we don't expect

Reply:
I dig it -- it makes the ErrorBoundary more important (will prioritize as separate feature), but helps make our code more explicit